src/coreclr/gc/gc.cpp

```diff
@@ -10939,44 +10952,22 @@ heap_segment* gc_heap::get_free_region (int gen_number, size_t size)
     if (gen_number <= max_generation)
     {
         assert (size == 0);
-        if (free_regions)
-        {
-            num_free_regions--;
-            num_free_regions_removed++;
-            region = free_regions;
-            dprintf (REGIONS_LOG, ("%d free regions left, get %Ix-%Ix-%Ix",
-                num_free_regions, heap_segment_mem (region),
-                heap_segment_committed (region), heap_segment_used (region)));
-            free_regions = heap_segment_next (free_regions);
-            committed_in_free -= heap_segment_committed (region) - get_region_start (region);
-        }
+        region = free_regions[basic_free_region].unlink_region_front();
     }
     else
     {
-        // look for a region that is large enough
-        heap_segment** link = &free_large_regions;
-        for (region = free_large_regions; region != nullptr; region = heap_segment_next (region))
-        {
-            uint8_t* region_start = get_region_start(region);
-            uint8_t* region_end = heap_segment_reserved(region);
+        const size_t LARGE_REGION_SIZE = global_region_allocator.get_large_region_alignment();
 
-            if ((region_end - region_start) >= (ptrdiff_t)size)
-            {
-                // found a region that is large enough
-                break;
-            }
-            link = &heap_segment_next (region);
-        }
+        free_region_kind kind = size <= LARGE_REGION_SIZE ? large_free_region : huge_free_region;
+
+        // get it from the local list of free large regions if possible
+        region = free_regions[kind].unlink_region_front();
 
-        if (region != nullptr)
+        if (region == nullptr)
         {
-            num_free_large_regions--;
-            num_free_large_regions_removed++;
-            dprintf (REGIONS_LOG, ("%d large free regions left, get %Ix-%Ix-%Ix",
-                num_free_large_regions, heap_segment_mem (region),
-                heap_segment_committed (region), heap_segment_used (region)));
-            *link = heap_segment_next (region);
-            committed_in_free -= heap_segment_committed (region) - get_region_start (region);
+            ASSERT_HOLDING_SPIN_LOCK (&gc_lock);
+
+            region = global_free_huge_regions.unlink_smallest_region (size);
```

Comment:
I'm worried about just removing the whole region here. imagine a scenario where we have a few huge regions and not many large regons and a few heaps will exhaust those few and other heaps end up not being able to get any

Reply:
I changed the code so that a request for a large region will not get a huge region from the free list. Perhaps it would be better to split the region in this case, but that seems a bit complicated.

Reply:
I believe we'll definitely need to split up regions but that doens't need to be done in this PR.